baseplate/context/memcache/lib.py

```diff
@@ -0,0 +1,226 @@
+
+"""
+
+
+class Flags(object):
+    """Memcached client flags
+
+    Flags are an arbitrary 16-bit unsigned integer that the memcache server
+    stores along with the data and sends back when the item is retrieved.
+    Clients may use this as a bit field to store data-specific information;
+    this field is opaque to the server.
+
+    """
+
+    JSON = 1 << 0
+    INTEGER = 1 << 1
+    LONG = 1 << 2
+    ZLIB = 1 << 3
+
+
+def decompress_and_load(key, serialized, flags):
+    """Deserialization method compatible with dump_and_compress().
+
+    :param str key: the memcached key.
+    :param str serialized: the serialized object returned from memcached.
+    :param int flags: value stored and returned from memcached for the client
+        to use to indicate how the value was serialized.
+    :returns str value: the deserialized value.
+
+    """
+
+    if flags & Flags.ZLIB:
+        serialized = zlib.decompress(serialized)
+        flags ^= Flags.ZLIB
+
+    if flags == 0:
+        return serialized
+    elif flags == Flags.INTEGER:
+        # python3 doesn't have a long integer type, so all integers are written
+        # with Flags.INTEGER. This means that a value written by a python3
+        # client can exceed the maximum integer value (sys.maxint). This
+        # appears to be ok--python2 will automatically convert to long if the
+        # value is too large.
+        return int(serialized)
+    elif flags == Flags.LONG:
+        return long(serialized)
+    else:
+        try:
+            return json.loads(serialized)
+        except Exception:
+            logging.info('json error', exc_info=True)
+            return None
+
+
+def make_dump_and_compress_fn(min_compress_length=0, compress_level=1):
+    """Create a serialization method compatible with decompress_and_load().
+
+    The resulting method is a chain of json.loads and zlib compression.
+
+    :param int min_compress_length: the minimum serialized string length to
+        enable zlib compression. 0 disables compression.
+    :param int compress_level: zlib compression level. 0 disables compression
+        and 9 is the maximum value.
+    :returns func memcache_serializer: the serializer method.
+
+    """
+
+    assert min_compress_length >= 0
+    assert 0 <= compress_level <= 9
+
+    def dump_and_compress(key, value):
+        """Serialization method compatible with decompress_and_load().
+
+        :param str key: the memcached key.
+        :param value: python object to be serialized and set to memcached.
+        :returns: value serialized as str, flags int.
+
+        """
+
+        if isinstance(value, string_types):
+            serialized = value
+            flags = 0
+        elif isinstance(value, int):
+            serialized = "%d" % value
+            flags = Flags.INTEGER
+        elif isinstance(value, long):
+            serialized = "%d" % value
+            flags = Flags.LONG
+        else:
+            serialized = json.dumps(value)
```

Comment:
This can throw `TypeError` if value isn't serializable to JSON, I think expecting callers to handle this is totally reasonable however maybe it should be noted in docstring above that they are expected to possibly handle those errors